### ISSUE TITLE
[JENKINS-45688] Unable to set extensions for Git SCM Source via DSL

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#3153]
<p>In the past I used `gitSCMSource` to specify my git configuration with extensions in my Job DSL.<br/>
 <br/>
Template I used:</p>

<pre>
multibranchPipelineJob(<span class="code-quote">'${JOB_NAME}'</span>) {
    displayName(<span class="code-quote">'${JOB_NAME}'</span>)
    description(<span class="code-quote">'${JOB_DESCRIPTION}'</span>)
    orphanedItemStrategy {
        discardOldItems {
            numToKeep(7)
        }
    }
    branchSources {
        branchSource {
            source {
                gitSCMSource {
                    id('')
                    remoteName(<span class="code-quote">'origin'</span>)
                    rawRefSpecs('')
                    ignoreOnPushNotifications(<span class="code-keyword">false</span>)
                    remote(<span class="code-quote">'${JOB_GIT_URL}'</span>)
                    credentialsId(<span class="code-quote">'${GIT_CREDENTIALS_ID}'</span>)
                    includes(<span class="code-quote">'*'</span>)
                    excludes('')
                    extensions {
                        wipeWorkspace()
                        preBuildMerge {
                            options {
                                mergeRemote(<span class="code-quote">'origin'</span>)
                                mergeTarget(<span class="code-quote">'${GIT_BRANCH}'</span>)
                                mergeStrategy(<span class="code-quote">'<span class="code-keyword">default</span>'</span>)
                                fastForwardMode(<span class="code-quote">'FF'</span>)
                            }
                        }
                    }
                }
            }
        }
    }
}</pre>

<p>Now after updating to latest Jenkins/Plugins this no longer works. I looked at using the `git` source instead of `gitSCMSource` but it does not provide a way to set the extensions.</p>

<p>In the new git-plugin 3.4.0 extensions are deprecated and are now traits. Not sure if this has anything to do with the issue.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=arty13">arty13</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688" target="_blank">Unable to set extensions for Git SCM Source via DSL</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: git-plugin, job-dsl-plugin, workflow-multibranch-plugin
<li><b>label(s)</b>: git-plugin, job-dsl, plugin
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2017-08-29T10:41:09+00:00
<li><b>votes</b>: 1
<li><b>watchers</b>: 9
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>In the past I used `gitSCMSource` to specify my git configuration with extensions in my Job DSL.
 
Template I used:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">multibranchPipelineJob(<span class="code-quote">'${JOB_NAME}'</span>) {
    displayName(<span class="code-quote">'${JOB_NAME}'</span>)
    description(<span class="code-quote">'${JOB_DESCRIPTION}'</span>)
    orphanedItemStrategy {
        discardOldItems {
            numToKeep(7)
        }
    }
    branchSources {
        branchSource {
            source {
                gitSCMSource {
                    id('')
                    remoteName(<span class="code-quote">'origin'</span>)
                    rawRefSpecs('')
                    ignoreOnPushNotifications(<span class="code-keyword">false</span>)
                    remote(<span class="code-quote">'${JOB_GIT_URL}'</span>)
                    credentialsId(<span class="code-quote">'${GIT_CREDENTIALS_ID}'</span>)
                    includes(<span class="code-quote">'*'</span>)
                    excludes('')
                    extensions {
                        wipeWorkspace()
                        preBuildMerge {
                            options {
                                mergeRemote(<span class="code-quote">'origin'</span>)
                                mergeTarget(<span class="code-quote">'${GIT_BRANCH}'</span>)
                                mergeStrategy(<span class="code-quote">'<span class="code-keyword">default</span>'</span>)
                                fastForwardMode(<span class="code-quote">'FF'</span>)
                            }
                        }
                    }
                }
            }
        }
    }
}</pre>
</div></div>

<p>Now after updating to latest Jenkins/Plugins this no longer works. I looked at using the `git` source instead of `gitSCMSource` but it does not provide a way to set the extensions.</p>

<p>In the new git-plugin 3.4.0 extensions are deprecated and are now traits. Not sure if this has anything to do with the issue.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins: 2.60.2<br/>
workflow-multibranch-plugin: 1.16<br/>
git-plugin: 3.4.1<br/>
job-dsl-plugin: 1.64
```
</details>
<details><summary><i>2 attachments</i></summary>

- [Screen Shot 2017-07-21 at 09.36.06.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38989/Screen%20Shot%202017-07-21%20at%2009.36.06.png)
  > ![Screen Shot 2017-07-21 at 09.36.06.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/38989/Screen%20Shot%202017-07-21%20at%2009.36.06.png)
- [Screen Shot 2017-07-21 at 09.36.26.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38988/Screen%20Shot%202017-07-21%20at%2009.36.26.png)
  > ![Screen Shot 2017-07-21 at 09.36.26.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/38988/Screen%20Shot%202017-07-21%20at%2009.36.26.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=307773&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307773">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I suspect the fact that it worked before was not an intended use of the GitSCMSource API.  <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenc" class="user-hover" rel="stephenc">stephenc</a> will have more insights on the expectation of the use of the extensions from GitSCMSource.</p>

<p>Does the technique in the <a href="https://jenkinsci.github.io/job-dsl-plugin/#path/job-scm-git" class="external-link" target="_blank" rel="nofollow noopener">official documentation</a> work?  It seems to be setting extensions using a git node in the DSL, without using GitSCMSource.</p>

<p>Does the technique described in "<a href="http://www.devexp.eu/2014/10/26/use-unsupported-jenkins-plugins-with-jenkins-dsl/" class="external-link" target="_blank" rel="nofollow noopener">use unsupported plugins with job DSL</a>" still work?  That technique does not seem to require GitSCMSource.</p>

<p>A similar technique is described in "<a href="https://stackoverflow.com/questions/27929548/how-to-use-jenkins-jobdsl-to-set-check-out-to-specific-local-branch-in-git-plu" class="external-link" target="_blank" rel="nofollow noopener">How to use Jenkins JobDSL to set 'Check out to specific local branch' in Git Plugin</a>" , in "<a href="https://stackoverflow.com/questions/25943651/sparse-checkout-with-jenkins-dsl-plugin" class="external-link" target="_blank" rel="nofollow noopener">sparse checkout with jenkins dsl plugin</a>" and in "<a href="https://stackoverflow.com/questions/41462362/use-jenkins-dsl-to-specify-a-a-git-executable-in-a-github-scm-node" class="external-link" target="_blank" rel="nofollow noopener">use Jenkins DSL to specify a git executable in github SCM</a>".</p>

<p>I'm prone to close this as "Won't fix", but will wait for verification from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a></p>
</pre>
</details>

</ul>
</details>

<p>I suspect the fact that it worked before was not an intended use of the GitSCMSource API.  <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenc" class="user-hover" rel="stephenc">stephenc</a> will have more insights on the expectation of the use of the extensions from GitSCMSource.</p>

<p>Does the technique in the <a href="https://jenkinsci.github.io/job-dsl-plugin/#path/job-scm-git" class="external-link" target="_blank" rel="nofollow noopener">official documentation</a> work?  It seems to be setting extensions using a git node in the DSL, without using GitSCMSource.</p>

<p>Does the technique described in "<a href="http://www.devexp.eu/2014/10/26/use-unsupported-jenkins-plugins-with-jenkins-dsl/" class="external-link" target="_blank" rel="nofollow noopener">use unsupported plugins with job DSL</a>" still work?  That technique does not seem to require GitSCMSource.</p>

<p>A similar technique is described in "<a href="https://stackoverflow.com/questions/27929548/how-to-use-jenkins-jobdsl-to-set-check-out-to-specific-local-branch-in-git-plu" class="external-link" target="_blank" rel="nofollow noopener">How to use Jenkins JobDSL to set 'Check out to specific local branch' in Git Plugin</a>" , in "<a href="https://stackoverflow.com/questions/25943651/sparse-checkout-with-jenkins-dsl-plugin" class="external-link" target="_blank" rel="nofollow noopener">sparse checkout with jenkins dsl plugin</a>" and in "<a href="https://stackoverflow.com/questions/41462362/use-jenkins-dsl-to-specify-a-a-git-executable-in-a-github-scm-node" class="external-link" target="_blank" rel="nofollow noopener">use Jenkins DSL to specify a git executable in github SCM</a>".</p>

<p>I'm prone to close this as "Won't fix", but will wait for verification from <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenconnolly" class="user-hover" rel="stephenconnolly">stephenconnolly</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stephenconnolly.png" title="stephenconnolly's avatar" /> stephenconnolly:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=307779&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307779">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>So my understanding is that you need to use the techniques for using unsupported plugins with job DSL to configure traits... at least until we get around to adding the <tt>@Symbol</tt> annotations (<a href="https://issues.jenkins.io/browse/JENKINS-45504" title="Add @Symbol annotations to traits" class="issue-link" data-issue-key="JENKINS-45504"><del>JENKINS-45504</del></a>)... of course that is currently blocked on <a href="https://issues.jenkins.io/browse/JENKINS-45503" title="Consolidate common traits into SCM-API" class="issue-link" data-issue-key="JENKINS-45503">JENKINS-45503</a> because for reasons documented in those issues.</p>

<p>Having said all that, we did retain the setter for extensions (just marked deprecated is all) and that setter will create the matching traits, so the only thing I can think is that you may be hitting the ref trimming... in which case all you need to do is add the trait that allows specifying an additional refspec to fetch and ensure that the master branch refspec is always fetched from the remote so that the merge extension / trait can work...</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> I think we need to hear back on how this "no longer works" to determine if:</p>

<ul>
	<li>The jobs are not getting configured correctly; or</li>
	<li>The configured jobs are failing because of the missing refspec for origin; or</li>
	<li>Something else</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>So my understanding is that you need to use the techniques for using unsupported plugins with job DSL to configure traits... at least until we get around to adding the <tt>@​Symbol</tt> annotations (<a href="https://issues.jenkins.io/browse/JENKINS-45504" title="Add @​Symbol annotations to traits" class="issue-link" data-issue-key="JENKINS-45504"><del>JENKINS-45504</del></a>)... of course that is currently blocked on <a href="https://issues.jenkins.io/browse/JENKINS-45503" title="Consolidate common traits into SCM-API" class="issue-link" data-issue-key="JENKINS-45503">JENKINS-45503</a> because for reasons documented in those issues.</p>

<p>Having said all that, we did retain the setter for extensions (just marked deprecated is all) and that setter will create the matching traits, so the only thing I can think is that you may be hitting the ref trimming... in which case all you need to do is add the trait that allows specifying an additional refspec to fetch and ensure that the master branch refspec is always fetched from the remote so that the merge extension / trait can work...</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> I think we need to hear back on how this "no longer works" to determine if:</p>

<ul>
	<li>The jobs are not getting configured correctly; or</li>
	<li>The configured jobs are failing because of the missing refspec for origin; or</li>
	<li>Something else</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>arty13:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=307849&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307849">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Using this example:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">multibranchPipelineJob(<span class="code-quote">'AMQ-Docker-Branch'</span>) {
  displayName(<span class="code-quote">'AMQ-Docker-Branch'</span>)
  description(<span class="code-quote">'AMQ Docker Container Branch Builds'</span>)
  orphanedItemStrategy {
    discardOldItems {
      numToKeep(7)
    }
  }
  branchSources {
    git {
      id('')
      ignoreOnPushNotifications(<span class="code-keyword">false</span>)
      remote(<span class="code-quote">'git@git.com/amq-docker.git'</span>)
      credentialsId(<span class="code-quote">'SSHKeyCredentials'</span>)
      includes(<span class="code-quote">'*'</span>)
      excludes(<span class="code-quote">'master'</span>)
      extensions {
        wipeWorkspace()
        preBuildMerge {
          options {
            mergeRemote(<span class="code-quote">'origin'</span>)
            mergeTarget(<span class="code-quote">'master'</span>)
            mergeStrategy(<span class="code-quote">'<span class="code-keyword">default</span>'</span>)
            fastForwardMode(<span class="code-quote">'FF'</span>)
          }
        }
      }
    }
  }
}
</pre>
</div></div>

<p>I get this error:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">ERROR: (script, line 17) No signature of method: javaposse.jobdsl.dsl.helpers.workflow.GitBranchSourceContext.extensions() is applicable <span class="code-keyword">for</span> argument types: (script$_run_closure1$_closure3$_closure5$_closure6) values: [script$_run_closure1$_closure3$_closure5$_closure6@5218bc64]
Finished: FAILURE
</pre>
</div></div>

<p>If I try to do the unsupported approach - I am using this sample, but it does not create the extensions.</p>


<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">multibranchPipelineJob(<span class="code-quote">'CI-TEST-MB'</span>) {
  displayName(<span class="code-quote">'CI-TEST-MB'</span>)
  description(<span class="code-quote">'Test Multibranch Job'</span>)
  orphanedItemStrategy {
    discardOldItems {
      numToKeep(7)
    }
  }
branchSources {
  git {
    id('')
    ignoreOnPushNotifications(<span class="code-keyword">false</span>)
    remote(<span class="code-quote">'git@git.com/test.git'</span>)
    credentialsId(<span class="code-quote">'test'</span>)
    includes(<span class="code-quote">'*'</span>)
    excludes(<span class="code-quote">'master'</span>)
    configure { node -&gt; <span class="code-comment">//the GitSCM node is passed in
</span>
        <span class="code-comment">// Add the WipeWorkspace functionality
</span>        node / <span class="code-quote">'extensions'</span> &lt;&lt; <span class="code-quote">'hudson.plugins.git.extensions.impl.WipeWorkspace'</span>  {
        }
      }
    }
  }
}
</pre>
</div></div>
<p> See attachements for result..</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Using this example:</p>

<pre>
multibranchPipelineJob(<span class="code-quote">'AMQ-Docker-Branch'</span>) {
  displayName(<span class="code-quote">'AMQ-Docker-Branch'</span>)
  description(<span class="code-quote">'AMQ Docker Container Branch Builds'</span>)
  orphanedItemStrategy {
    discardOldItems {
      numToKeep(7)
    }
  }
  branchSources {
    git {
      id('')
      ignoreOnPushNotifications(<span class="code-keyword">false</span>)
      remote(<span class="code-quote">'git@​git.com/amq-docker.git'</span>)
      credentialsId(<span class="code-quote">'SSHKeyCredentials'</span>)
      includes(<span class="code-quote">'*'</span>)
      excludes(<span class="code-quote">'master'</span>)
      extensions {
wipeWorkspace()
preBuildMerge {
  options {
    mergeRemote(<span class="code-quote">'origin'</span>)
    mergeTarget(<span class="code-quote">'master'</span>)
    mergeStrategy(<span class="code-quote">'<span class="code-keyword">default</span>'</span>)
    fastForwardMode(<span class="code-quote">'FF'</span>)
  }
}
      }
    }
  }
}
</pre>

<p>I get this error:</p>

<pre>
ERROR: (script, line 17) No signature of method: javaposse.jobdsl.dsl.helpers.workflow.GitBranchSourceContext.extensions() is applicable <span class="code-keyword">for</span> argument types: (script$_run_closure1$_closure3$_closure5$_closure6) values: [script$_run_closure1$_closure3$_closure5$_closure6@​5218bc64]
Finished: FAILURE
</pre>

<p>If I try to do the unsupported approach - I am using this sample, but it does not create the extensions.</p>



<pre>
multibranchPipelineJob(<span class="code-quote">'CI-TEST-MB'</span>) {
  displayName(<span class="code-quote">'CI-TEST-MB'</span>)
  description(<span class="code-quote">'Test Multibranch Job'</span>)
  orphanedItemStrategy {
    discardOldItems {
      numToKeep(7)
    }
  }
branchSources {
  git {
    id('')
    ignoreOnPushNotifications(<span class="code-keyword">false</span>)
    remote(<span class="code-quote">'git@​git.com/test.git'</span>)
    credentialsId(<span class="code-quote">'test'</span>)
    includes(<span class="code-quote">'*'</span>)
    excludes(<span class="code-quote">'master'</span>)
    configure { node -> <span class="code-comment">//the GitSCM node is passed in
</span>
<span class="code-comment">// Add the WipeWorkspace functionality
</span>node / <span class="code-quote">'extensions'</span> << <span class="code-quote">'hudson.plugins.git.extensions.impl.WipeWorkspace'</span>  {
}
      }
    }
  }
}
</pre>
<p> See attachements for result..</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>arty13:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=307855&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307855">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>With my latest testing using the configure block - I noticed the node reference is always referencing the root xml node, not the current node. I will continue to test/configure using the job dsl config block and will let you know what I find.</p>
</pre>
</details>

</ul>
</details>

<p>With my latest testing using the configure block - I noticed the node reference is always referencing the root xml node, not the current node. I will continue to test/configure using the job dsl config block and will let you know what I find.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>arty13:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=307864&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-307864">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>A hackish workaround using the configure block (bypassing git block entirely)..</p>


<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">multibranchPipelineJob(<span class="code-quote">'CI-TEST-MB'</span>) {
  displayName(<span class="code-quote">'CI-TEST-MB'</span>)
  description(<span class="code-quote">'Test Multibranch Job'</span>)
  orphanedItemStrategy {
    discardOldItems {
      numToKeep(7)
    }
  }
  branchSources {
    configure { node -&gt;
      node / sources(class: <span class="code-quote">'jenkins.branch.MultiBranchProject$BranchSourceList'</span>) / data / <span class="code-quote">'jenkins.branch.BranchSource'</span> / source(class: <span class="code-quote">'jenkins.plugins.git.GitSCMSource'</span>) {
        remote <span class="code-quote">'git@git.com/test.git'</span>
        credentialsId <span class="code-quote">'test'</span>
        includes <span class="code-quote">'*'</span>
        excludes <span class="code-quote">'master'</span>
        ignoreOnPushNotifications <span class="code-quote">'<span class="code-keyword">false</span>'</span>
        id ''
        traits {
          <span class="code-quote">'jenkins.plugins.git.traits.BranchDiscoveryTrait'</span>()
          <span class="code-quote">'jenkins.plugins.git.traits.WipeWorkspaceTrait'</span>() {
            extension(class: <span class="code-quote">'hudson.plugins.git.extensions.impl.WipeWorkspace'</span>)
          }
          <span class="code-quote">'jenkins.plugins.git.traits.PreBuildMergeTrait'</span>() {
            extension(class: <span class="code-quote">'hudson.plugins.git.extensions.impl.PreBuildMerge'</span>) {
              options {
                mergeRemote <span class="code-quote">'origin'</span>
                mergeTarget <span class="code-quote">'master'</span>
                mergeStrategy <span class="code-quote">'<span class="code-keyword">default</span>'</span>
                fastForwardMode <span class="code-quote">'FF'</span>
              }
            }
          }
        }
      }
    }
  }
}
</pre>
</div></div>
<p> </p>

<p>Reason I removed the git {} section entirely was because it would always add a new record of the 'jenkins.branch.BranchSource' section (not updating the existing BranchSource node)</p>

<p>This might be a separate bug (or misunderstood documentation) If I put the configure block inside the git{} section, the reference 'node' would always point to the root xml element not the current node/git element. 
Example from documentation..</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">job(<span class="code-quote">'example'</span>) {
    scm {
        git {
            remote {
                url(<span class="code-quote">'git@server:account/repo1.git'</span>)
            }
            configure { node -&gt;
                <span class="code-comment">// node represents &lt;hudson.plugins.git.GitSCM&gt;
</span>            }
        }
    }
}
</pre>
</div></div>
<p> </p>
</pre>
</details>

</ul>
</details>

<p>A hackish workaround using the configure block (bypassing git block entirely)..</p>



<pre>
multibranchPipelineJob(<span class="code-quote">'CI-TEST-MB'</span>) {
  displayName(<span class="code-quote">'CI-TEST-MB'</span>)
  description(<span class="code-quote">'Test Multibranch Job'</span>)
  orphanedItemStrategy {
    discardOldItems {
      numToKeep(7)
    }
  }
  branchSources {
    configure { node ->
      node / sources(class: <span class="code-quote">'jenkins.branch.MultiBranchProject$BranchSourceList'</span>) / data / <span class="code-quote">'jenkins.branch.BranchSource'</span> / source(class: <span class="code-quote">'jenkins.plugins.git.GitSCMSource'</span>) {
remote <span class="code-quote">'git@​git.com/test.git'</span>
credentialsId <span class="code-quote">'test'</span>
includes <span class="code-quote">'*'</span>
excludes <span class="code-quote">'master'</span>
ignoreOnPushNotifications <span class="code-quote">'<span class="code-keyword">false</span>'</span>
id ''
traits {
  <span class="code-quote">'jenkins.plugins.git.traits.BranchDiscoveryTrait'</span>()
  <span class="code-quote">'jenkins.plugins.git.traits.WipeWorkspaceTrait'</span>() {
    extension(class: <span class="code-quote">'hudson.plugins.git.extensions.impl.WipeWorkspace'</span>)
  }
  <span class="code-quote">'jenkins.plugins.git.traits.PreBuildMergeTrait'</span>() {
    extension(class: <span class="code-quote">'hudson.plugins.git.extensions.impl.PreBuildMerge'</span>) {
      options {
mergeRemote <span class="code-quote">'origin'</span>
mergeTarget <span class="code-quote">'master'</span>
mergeStrategy <span class="code-quote">'<span class="code-keyword">default</span>'</span>
fastForwardMode <span class="code-quote">'FF'</span>
      }
    }
  }
}
      }
    }
  }
}
</pre>
<p> </p>

<p>Reason I removed the git {} section entirely was because it would always add a new record of the 'jenkins.branch.BranchSource' section (not updating the existing BranchSource node)</p>

<p>This might be a separate bug (or misunderstood documentation) If I put the configure block inside the git{} section, the reference 'node' would always point to the root xml element not the current node/git element. <br/>
Example from documentation..</p>

<pre>
job(<span class="code-quote">'example'</span>) {
    scm {
git {
    remote {
url(<span class="code-quote">'git@​server:account/repo1.git'</span>)
    }
    configure { node ->
<span class="code-comment">// node represents <hudson.plugins.git.GitSCM>
</span>    }
}
    }
}
</pre>
<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/daspilker.jpg" title="daspilker's avatar" /> daspilker:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=311890&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-311890">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I can not reproduce this with all plugins updated to the latest version. I see <em>multibranchPipelineJob &gt; branchSources &gt; branchSource &gt; source &gt; git &gt; extensions</em> as well as <em>multibranchPipelineJob &gt; branchSources &gt; branchSource &gt; source &gt; git &gt; traits</em> in my embedded API viewer. Can you create a working configuration manually and post the config XML? Then we can figure out what's missing in Job DSL.</p>
</pre>
</details>

</ul>
</details>

<p>I can not reproduce this with all plugins updated to the latest version. I see <em>multibranchPipelineJob > branchSources > branchSource > source > git > extensions</em> as well as <em>multibranchPipelineJob > branchSources > branchSource > source > git > traits</em> in my embedded API viewer. Can you create a working configuration manually and post the config XML? Then we can figure out what's missing in Job DSL.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/sag47.jpg" title="sag47's avatar" /> sag47:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338459&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338459">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=daspilker" class="user-hover" rel="daspilker">daspilker</a> I encountered this issue after installing blue ocean.  Blue ocean (for some strange reason) requires bit bucket branch source and github branch source plugins to operate.  With those plugins installed one faces the issue when trying to configure the git plugin as a branch source.</p>

<p>When I manually disable those two plugins (breaking blue ocean) I get a working DSL again.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=daspilker" class="user-hover" rel="daspilker">daspilker</a> I encountered this issue after installing blue ocean.  Blue ocean (for some strange reason) requires bit bucket branch source and github branch source plugins to operate.  With those plugins installed one faces the issue when trying to configure the git plugin as a branch source.</p>

<p>When I manually disable those two plugins (breaking blue ocean) I get a working DSL again.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/sag47.jpg" title="sag47's avatar" /> sag47:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338543&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338543">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Here's a link to the DSL used.</p>

<p><a href="https://github.com/gimp-ci/jenkins-dsl/blob/292cafd41e8c26e200dbe967aba7ec71e120df84/jobs/gimp_multibranch_pipelines.groovy" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/gimp-ci/jenkins-dsl/blob/292cafd41e8c26e200dbe967aba7ec71e120df84/jobs/gimp_multibranch_pipelines.groovy</a></p>

<p>Here's a list of plugins installed that broke the the DSL.  <a href="https://github.com/gimp-ci/jenkins-os-packages/commit/2e9e0e6b5a2e0287268ca70d836f2a7e71fef72b" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/gimp-ci/jenkins-os-packages/commit/2e9e0e6b5a2e0287268ca70d836f2a7e71fef72b</a></p>

<p>In particular I confirmed the following plugins broke the DSL.</p>

<ul>
	<li>org.jenkins-ci.plugins:cloudbees-bitbucket-branch-source:2.2.11</li>
	<li>org.jenkins-ci.plugins:github-branch-source:2.3.4</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>Here's a link to the DSL used.</p>

<p><a href="https://github.com/gimp-ci/jenkins-dsl/blob/292cafd41e8c26e200dbe967aba7ec71e120df84/jobs/gimp_multibranch_pipelines.groovy" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/gimp-ci/jenkins-dsl/blob/292cafd41e8c26e200dbe967aba7ec71e120df84/jobs/gimp_multibranch_pipelines.groovy</a></p>

<p>Here's a list of plugins installed that broke the the DSL.  <a href="https://github.com/gimp-ci/jenkins-os-packages/commit/2e9e0e6b5a2e0287268ca70d836f2a7e71fef72b" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/gimp-ci/jenkins-os-packages/commit/2e9e0e6b5a2e0287268ca70d836f2a7e71fef72b</a></p>

<p>In particular I confirmed the following plugins broke the DSL.</p>

<ul>
	<li>org.jenkins-ci.plugins:cloudbees-bitbucket-branch-source:2.2.11</li>
	<li>org.jenkins-ci.plugins:github-branch-source:2.3.4</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/sag47.jpg" title="sag47's avatar" /> sag47:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338545&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338545">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This nonsense requiring the configure block is ridiculous.  I've spent the past 5 hours trying to figure it out and I'm definitely not a Job DSL newb.</p>
</pre>
</details>

</ul>
</details>

<p>This nonsense requiring the configure block is ridiculous.  I've spent the past 5 hours trying to figure it out and I'm definitely not a Job DSL newb.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>stephenking:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338549&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338549">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sag47" class="user-hover" rel="sag47">sag47</a> I am sure this is not by intention and the maintainers will welcome your pull requests to stop this nonsense.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sag47" class="user-hover" rel="sag47">sag47</a> I am sure this is not by intention and the maintainers will welcome your pull requests to stop this nonsense.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/sag47.jpg" title="sag47's avatar" /> sag47:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338551&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338551">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenking" class="user-hover" rel="stephenking">stephenking</a> my comments have added a lot of value to this issue by referencing both source code, examples, and reproduction.  Your comment claims I should stop doing this.  I disagree.</p>

<p>Edit: I think I initially misread your comment.  In any case, "contribute code" should never be the default retort as a comeback in my opinion.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=stephenking" class="user-hover" rel="stephenking">stephenking</a> my comments have added a lot of value to this issue by referencing both source code, examples, and reproduction.  Your comment claims I should stop doing this.  I disagree.</p>

<p>Edit: I think I initially misread your comment.  In any case, "contribute code" should never be the default retort as a comeback in my opinion.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>stephenking:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338553&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338553">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>No, my point was "please be friendly". Nobody creates "nonsense" by intention.</p>
</pre>
</details>

</ul>
</details>

<p>No, my point was "please be friendly". Nobody creates "nonsense" by intention.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/sag47.jpg" title="sag47's avatar" /> sag47:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338555&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338555">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Sure, I can put my frustration in check.  My intent wasn't to be hostile nor to imply it was done on purpose (it would be odd to assume that it's on purpose).</p>
</pre>
</details>

</ul>
</details>

<p>Sure, I can put my frustration in check.  My intent wasn't to be hostile nor to imply it was done on purpose (it would be odd to assume that it's on purpose).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/sag47.jpg" title="sag47's avatar" /> sag47:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338556&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338556">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Added git-plugin since the git-plugin should define "@Symbol" for its classes to help with the DSL generation to be unique.</p>
</pre>
</details>

</ul>
</details>

<p>Added git-plugin since the git-plugin should define "@​Symbol" for its classes to help with the DSL generation to be unique.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/sag47.jpg" title="sag47's avatar" /> sag47:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338565&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338565">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://github.com/jenkinsci/git-plugin/pull/595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-plugin/pull/595</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://github.com/jenkinsci/git-plugin/pull/595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/git-plugin/pull/595</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>stephenking:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338592&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338592">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sag47" class="user-hover" rel="sag47">sag47</a>, thanks for bringing this forward!</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=sag47" class="user-hover" rel="sag47">sag47</a>, thanks for bringing this forward!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/sag47.jpg" title="sag47's avatar" /> sag47:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-45688?focusedId=338679&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-338679">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>You're welcome.</p>
</pre>
</details>

</ul>
</details>

<p>You're welcome.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/git-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-45504%22">JENKINS-45504</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/git-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-46202%22">JENKINS-46202</a>]</i>
